### ISSUE TITLE
Discontinuities when simulating periodic signals in neurodsp.sim

### ISSUE THREAD

**Issue by nschawor** [neurodsp-tools/neurodsp#174]
Stationary simulated periodic signals display some discontinuities.
<img width="676" alt="Screen Shot 2019-12-17 at 10 23 04" src="https://user-images.githubusercontent.com/1282430/72542834-a02fcd00-3839-11ea-92a7-e01c79331775.png">

Using the first three blocks of example code from here: https://neurodsp-tools.github.io/neurodsp/auto_tutorials/sim/plot_SimulatePeriodic.html

**Comment by TomDonoghue** [comment]
Okay, so this has been bugging me, and I wasn't sure if there was anything we could do to fix it, within the framework of simulating periodic components as concatenations of cycles (as we do).

This mismatch arises from uneven sampling, as interactions between the requested sampling rate and frequency of the cycle. If the signal sample rate doesn't play nice with the frequency, the cycles don't concatenate well. 

So, my idea / suggestion is to sample cycles at the nice sampling rate for them, build the signal, and then resample the whole signal to the desired sampling rate at the end. 

So, something like:
```
def sim_oscillation_test(n_seconds, fs, freq, cycle='sine', **cycle_params):

    # Figure out how many cycles are needed for the signal, & length of each cycle
    n_cycles = int(np.ceil(n_seconds * freq))
    # UPDATE: Change how `n_seconds_cycle` is computed (old version was weird)
    n_seconds_cycle = 1/freq 

    # UPDATE: Tune sampling rate for simulating cycle
    # Create oscillation by tiling a single cycle of the desired oscillation
    #   The cycle is created a sampling rate that is a multiple of the frequency
    #   This is done to create cycles that line up well for concatenation
    cycle_fs = freq * 100
    osc_cycle = sim_cycle(n_seconds_cycle, cycle_fs, cycle, **cycle_params)
    
    # Create full signal by tiling the simulated single cycle
    sig = np.tile(osc_cycle, n_cycles)
    
    # UPDATE: resample to request sampling rate
    # Resample cycle to desired sampling rate
    sig = resample(sig, int(n_cycles * 1/freq * fs))

    # Truncate the length of the signal to be the number of expected samples
    #   This is done because we simulate an integer number of cycles, 
    #   which can be beyond the requested data, given the freq & n_seconds input
    n_samps = int(n_seconds * fs)
    sig = sig[:n_samps]
    
    return sig
```

#### First pass tests
This seems to fix at least some, perhaps most of our problems. Most signals no longer show the temporal discontinuity, and lead to power spectra without the spiky-noise. 
There do however still seem to be some (not sure how many) edge cases that get the discontinuity. I'm not sure why but I think it's a floating point estimating thing somewhere.
I've explored so far with sine signals simulating continuous rhythms, but I think / hope this should play nicely with other cycles & burst signals. 

#### Question
Before I go too far down testing and trying this, and putting it in properly, I want to ask: does this make sense? Anyone have any insight into if this is likely to blow up somewhere and create more problems? If this doesn't work, I'm otherwise unsure how to make this cycle approach work out, and we might have to consider switching strategies altogether (which I would prefer not to do). 

Asking: @nschawor @rdgao @voytek and cc-ing @srcole (just in case you are bored and want to do any free consulting if you have any insight here :)).

**Comment by nschawor** [comment]
The two methods seem to slowly drift apart, from zero phase lag to increasing phase lag, is that right? Also not sure why the amplitude is different.

```python
fs = 1000
n_seconds = 10
osc_freq = 6.6

osc_sine = sim.sim_oscillation(n_seconds, fs, osc_freq, cycle='sine')
osc_sine2 = sim_oscillation_test(n_seconds, fs, osc_freq, cycle='sine')
times = create_times(n_seconds, fs)

fig, ax = plt.subplots()
ax.plot(times, osc_sine)
ax.plot(times, osc_sine2)
fig.show()
```

<img width="581" alt="Screen Shot 2020-02-12 at 23 25 41" src="https://user-images.githubusercontent.com/1282430/74410929-02570000-4def-11ea-8ae4-a2c8598ad96c.png">

**Comment by nschawor** [comment]
To keep it in writing and not forget, but no easy solution: In general, when cycles of different periods are not allowed with this cycle tiling approach, it represents a discontinuity in the conceptual model as in the spectral parametrization the oscillatory bumps are modelled by a Gaussian. If one would apply the same rationale here and tile periods sampled from a Gaussian distribution together, the spectral representation is polluted by these sampling artefacts.

So to have a sound unified conceptual model, one needs to find another way.

**Comment by TomDonoghue** [comment]
Okay, coming back around to this:

Good catch on the shift. It turns out though, that comparing to the current (or 'old') isn't really the best comparison though, since it turns out it's these 'old' simulations that have that lag, not the new ones (as can be seen by comparing to a non-tile comparison). 

The non-tile is a standard sine wave:
```
# Non-tiling
samples = np.linspace(0, n_seconds, int(n_seconds * fs), endpoint=True)
non_tile = np.sin(2 * np.pi * osc_freq * samples)
```

Compared to this, the old approach lags:
<img width="988" alt="Screen Shot 2020-06-15 at 2 11 42 AM" src="https://user-images.githubusercontent.com/7727566/84640131-2108f880-aeae-11ea-9aed-d471b9033b4b.png">

But the new approach does not:
<img width="997" alt="Screen Shot 2020-06-15 at 2 12 05 AM" src="https://user-images.githubusercontent.com/7727566/84640136-22d2bc00-aeae-11ea-86aa-4c45e04c4d23.png">

The difference in amplitude is just because there is normalization applied to the standard version in the module, which wasn't applied to the test function - that itself is not an issue. So, overall, in terms of lags (and getting rid of the discontinuities) we should be good with the new tiling approach.

**Comment by TomDonoghue** [comment]
For tracking purposes: this issue was explored in #189, which proposed the resamping suggested here, but it turns out it's the normalizations that created the discontinuities, and also resampling doesn't play well with variable cycle types, so that's not a general solution (nor is it needed). 

The fix for the normalizations was merged in #208.
